efel/tests/test_basic.py

```diff
@@ -2512,3 +2535,548 @@ def test_segfault_in_AP_begin_width():
+
+    # find start and end of decay
+    # has to be over deriv threshold for at least a given number of indices
+    pass_threshold_idxs = numpy.append(
+        -1, numpy.argwhere(dvdt > -min_derivative).flatten()
+    )
+    length_idx = numpy.argwhere(
+        numpy.diff(pass_threshold_idxs) > decay_start_min_length
+    )[0][0]
+    i_start = pass_threshold_idxs[length_idx] + 1
+
+    # find flat (end of decay)
+    flat_idxs = numpy.argwhere(dvdt[i_start:] > -min_derivative).flatten()
+    # for loop is not optimised
+    # but we expect the 1st few values to be the ones we are looking for
+    for i in flat_idxs:
+        i_flat = i + i_start
+        i_flat_stop = numpy.argwhere(
+            t_interval >= t_interval[i_flat] + t_length
+        )[0][0]
+        if numpy.mean(dvdt[i_flat:i_flat_stop]) > -min_derivative:
+            break
+
+    dvdt_decay = dvdt[i_start:i_flat]
+    t_decay = time[stim_start_idx + i_start:stim_start_idx + i_flat]
+    v_decay_tmp = voltage[stim_start_idx + i_start:stim_start_idx + i_flat]
+    v_decay = abs(v_decay_tmp - voltage[stim_start_idx + i_flat])
+
+    if len(dvdt_decay) < min_length:
+        return None
+
+    # -- golden search algorithm -- #
+    from scipy.optimize import minimize_scalar
+
+    def numpy_fit(x, t_decay, v_decay):
+        new_v_decay = v_decay + x
+        log_v_decay = numpy.log(new_v_decay)
+        (slope, _), res, _, _, _ = numpy.polyfit(
+            t_decay, log_v_decay, 1, full=True
+        )
+        range = numpy.max(log_v_decay) - numpy.min(log_v_decay)
+        return res / (range * range)
+
+    max_bound = min_derivative * 1000.
+    golden_bracket = [0, max_bound]
+    result = minimize_scalar(
+        numpy_fit,
+        args=(t_decay, v_decay),
+        bracket=golden_bracket,
+        method='golden',
+    )
+
+    # -- fit -- #
+    log_v_decay = numpy.log(v_decay + result.x)
+    slope, _ = numpy.polyfit(t_decay, log_v_decay, 1)
+
+    tau = -1. / slope
+    return abs(tau)
+
+
+def test_time_constant():
+    """basic: Test time constant with python implementation"""
+
+    import efel
+    efel.reset()
+
+    stim_start = 800.0
+    stim_end = 3800.0
+
+    time = efel.io.load_fragment('%s#col=1' % sagtrace1_url)
+    voltage = efel.io.load_fragment('%s#col=2' % sagtrace1_url)
+
+    trace = {}
+    trace['T'] = time
+    trace['V'] = voltage
+    trace['stim_start'] = [stim_start]
+    trace['stim_end'] = [stim_end]
+    time, voltage = interpolate(time, voltage, 0.1)
+
+    features = ["time_constant"]
+
+    feature_values = \
+        efel.getFeatureValues(
+            [trace],
+            features, raise_warnings=False)
+
+    time_cst = feature_values[0]["time_constant"]
+    assert len(time_cst) == 1
+
+    py_tau = py_time_constant(time, voltage, stim_start, stim_end)
+
+    # some difference because precision difference
+    # between efel and python implementation
+    numpy.testing.assert_allclose(time_cst, py_tau, rtol=1e-3)
```

Comment:
My curiosity: how much is the max absolute difference between python and C++ results?

Reply:
I have tried it on two traces: 
allfeatures/sagtrace_1.txt:
efel: 29.48777, python: 29.48297, absolute diff: 0.0048
basic/testdb1data.txt:
efel: 14.59336, python: 14.53353 absolute diff: 0.0598

Reply:
I see. Are you sure the difference is between the C++ and Python implementations? 0.0598 is not so small. Maybe there's a bug in the C++ implementation.

Reply:
Ah but probably the  `from scipy.optimize import minimize_scalar` implementation is different.

Reply:
The algorithm used in minimize_scalar is the same, but the precision is different. In efel, we stop the optimisation when we are at less than `0.01` from the right `x` value. So for `x` we have for efel `0.191403 `and for python `0.187260015`.

Note that here, `x` is not the `time constant`, but is needed to find the `time constant`